File: code/backend/core/src/storage/hash_index.cpp
#include "engine/storage/hash_index.hpp"

#include <algorithm>

namespace engine {

HashIndex::HashIndex(const HashIndex& other) : table(other.table), column(other.column) {
    std::lock_guard<std::mutex> g(other.mutex_);
    data_ = other.data_;
}

HashIndex& HashIndex::operator=(const HashIndex& other) {
    if (this != &other) {
        std::scoped_lock lock(mutex_, other.mutex_);
        table = other.table;
        column = other.column;
        data_ = other.data_;
    }
    return *this;
}

// mutex_ isn't movable -- manually moves the other members, leaving a fresh mutex_ in
// the moved-to object (same pattern LockManager/BPlusTree/QueryResultCache already use).
HashIndex::HashIndex(HashIndex&& other) noexcept : table(std::move(other.table)), column(std::move(other.column)) {
    std::lock_guard<std::mutex> g(other.mutex_);
    data_ = std::move(other.data_);
}

HashIndex& HashIndex::operator=(HashIndex&& other) noexcept {
    if (this == &other) return *this;
    std::scoped_lock lock(mutex_, other.mutex_);
    table = std::move(other.table);
    column = std::move(other.column);
    data_ = std::move(other.data_);
    return *this;
}

void HashIndex::rebuild(const std::vector<Row>& rows) {
    std::lock_guard<std::mutex> g(mutex_);
    data_.clear();
    for (const auto& row : rows) {
        auto it = row.find(column);
        if (it != row.end()) data_[it->second].push_back(row);
    }
}

std::vector<Row> HashIndex::get(const std::string& key) const {
    std::lock_guard<std::mutex> g(mutex_);
    auto it = data_.find(key);
    return it != data_.end() ? it->second : std::vector<Row>{};
}

void HashIndex::insert_row(const Row& row) {
    std::lock_guard<std::mutex> g(mutex_);
    auto it = row.find(column);
    if (it != row.end()) data_[it->second].push_back(row);
}

void HashIndex::remove_row(const std::string& col_val, const std::string& pk_col, const std::string& pk_val) {
    std::lock_guard<std::mutex> g(mutex_);
    auto it = data_.find(col_val);
    if (it == data_.end()) return;
    auto& bucket = it->second;
    bucket.erase(std::remove_if(bucket.begin(), bucket.end(),
                                 [&](const Row& r) {
                                     auto pit = r.find(pk_col);
                                     return pit != r.end() && pit->second == pk_val;
                                 }),
                 bucket.end());
    if (bucket.empty()) data_.erase(it);
}

std::size_t HashIndex::bucket_count() const {
    std::lock_guard<std::mutex> g(mutex_);
    return data_.size();
}

std::size_t HashIndex::row_count() const {
    std::lock_guard<std::mutex> g(mutex_);
    std::size_t total = 0;
    for (auto& [k, v] : data_) total += v.size();
    return total;
}

} // namespace engine

File: old/rusql-core/src/storage/hash_index.rs
use std::collections::HashMap;
use crate::engine::executor::Row;

/// 등호 조건 O(1) 검색을 위한 Hash Index.
/// key(컬럼 값) → 해당 행 목록을 HashMap으로 유지한다.
pub struct HashIndex {
    pub table:  String,
    pub column: String,
    data: HashMap<String, Vec<Row>>,
}

impl HashIndex {
    pub fn new(table: &str, column: &str) -> Self {
        Self { table: table.to_string(), column: column.to_string(), data: HashMap::new() }
    }

    /// 테이블 전체 행으로 인덱스를 (재)빌드한다.
    pub fn rebuild(&mut self, rows: &[Row]) {
        self.data.clear();
        for row in rows {
            if let Some(val) = row.get(&self.column) {
                self.data.entry(val.clone()).or_default().push(row.clone());
            }
        }
    }

    /// key에 해당하는 행 슬라이스를 반환한다.
    pub fn get(&self, key: &str) -> &[Row] {
        self.data.get(key).map(|v| v.as_slice()).unwrap_or(&[])
    }

    /// 단일 행을 인덱스에 추가한다 (O(1)).
    pub fn insert_row(&mut self, row: &Row) {
        if let Some(val) = row.get(&self.column) {
            self.data.entry(val.clone()).or_default().push(row.clone());
        }
    }

    /// PK 값이 `pk_val`인 행을 `col_val` 버킷에서 제거한다 (O(bucket size)).
    pub fn remove_row(&mut self, col_val: &str, pk_col: &str, pk_val: &str) {
        if let Some(bucket) = self.data.get_mut(col_val) {
            bucket.retain(|r| r.get(pk_col).map(|v| v != pk_val).unwrap_or(true));
            if bucket.is_empty() {
                self.data.remove(col_val);
            }
        }
    }

    /// 전체 버킷 수 (고유 키 수)
    pub fn bucket_count(&self) -> usize { self.data.len() }

    /// 전체 행 수
    pub fn row_count(&self) -> usize { self.data.values().map(|v| v.len()).sum() }
}
